### PR TITLE
Weighted LRU Cache

### PR DESCRIPTION
**Pull request by raj-prince**
### Description
1. This PR contains changes over [this](https://github.com/jacobsa/util/tree/master/lrucache) package
2. Added support for weighted entry, which can be used as normal lru cache by returning ValueType.Size() 1 every time.
3. Added support of mutex so that only one method of lrucache instance will run at a time.
4. Added unit test.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - Validated normal operation.
6. Unit tests - Automation
7. Integration tests - NA.

### CODE REVIEW

File: internal/cache/lrucache/lrucache_test.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+	defer c.Wrapped.CheckInvariants()
+
+	return c.Wrapped.LookUp(key)
+}
+
+////////////////////////////////////////////////////////////////////////
+// Boilerplate
+////////////////////////////////////////////////////////////////////////
+
+const maxWeight = 50
```
Comment (sethiay) [inline internal/cache/lrucache/lrucache_test.go]: let's keep this maxSize to be consistent with the maxSize in Cache.

File: internal/cache/lrucache/lrucache.go [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+	// Return the value.
+	return e.Value.(entry).Value
+}
+
+////////////////////////////////////////////////////////////////////////
+// Gob encoding
+////////////////////////////////////////////////////////////////////////
+
+func (c *Cache) GobEncode() (b []byte, err error) {
```
Comment (sethiay) [inline internal/cache/lrucache/lrucache.go:218]: Can we remove this to reduce code if not required ? We can include whenever there is a requirement.

File: go.mod [hunk 3]
```diff
@@ -23,8 +23,12 @@ require (
 	go.opencensus.io v0.24.0
 	golang.org/x/net v0.10.0
 	golang.org/x/oauth2 v0.8.0
+	golang.org/x/sync v0.2.0
+	golang.org/x/sys v0.8.0
+	golang.org/x/text v0.9.0
 	golang.org/x/time v0.3.0
 	google.golang.org/api v0.126.0
+	gopkg.in/yaml.v2 v2.4.0
```
Comment (sethiay) [inline go.mod:31]: Could you please help me understand why we are moving this from indirect to direct ?

**Review by sethiay — COMMENTED** [review]
Given this package will be used at multiple places, can we check we can write tests to check concurrency of exposed public methods ?

File: internal/cache/lrucache/lrucache.go [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+	return evictedValues
+}
+
+// Erase any entry for the supplied key, also returns the value of erased key.
+// LOCK_EXCLUDED(c.mu)
+func (c *Cache) Erase(key string) (value ValueType) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	e := c.index[key]
```
[the remark this answers is not in the document]
Comment (raj-prince) [reply internal/cache/lrucache/lrucache.go:180]: Both are good, but making in consistent. Using two value assignment at every places.

File: internal/cache/lrucache/lrucache.go [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+	// INVARIANT: Each element is of type entry
+	entries list.List
+
+	// Index of elements by name.
+	//
+	// INVARIANT: For each k, v: v.Value.(entry).Key == k
+	// INVARIANT: Contains all and only the elements of entries
+	index map[string]*list.Element
+
+	// All exposed methods are guarded by this Mutex. This means that only one
```
[the remark this answers is not in the document]
Comment (raj-prince) [reply internal/cache/lrucache/lrucache.go:59]: Done, with this change (updating -> accessing/updating)

File: internal/cache/lrucache/lrucache.go [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+		// Add the entry if already doesn't exist.
+		e := c.entries.PushFront(entry{key, value})
+		c.index[key] = e
+		c.currentSize += value.Size()
+	}
+
+	var evictedValues []ValueType
+	// Evict until we're at or below maxSize.
+	for c.currentSize > c.maxSize {
+		evictedValues = append(evictedValues, c.evictOne())
```
[the remark this answers is not in the document]
Comment (raj-prince) [reply internal/cache/lrucache/lrucache.go:168]: As discussed offline - Insertion in such cases will return a pre-defined error, which clients need to handle.

File: go.mod (hunk 3 above)
Comment (raj-prince) [reply go.mod:31]: We are using directly [here](https://github.com/GoogleCloudPlatform/gcsfuse/blob/66b8100310ff701a7f4fba5f801615c3480228dc/tools/integration_tests/util/setup/yaml-config.go#L9).

File: internal/cache/lrucache/lrucache.go [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+// all values are registered using gob.Register.
+//
+// May be used directly as a field in a larger struct. Must be created with New
+// or initialized using gob decoding.
+type Cache struct {
+	/////////////////////////
+	// Constant data
+	/////////////////////////
+
+	// INVARIANT: maxSize > 0
```
[the remark this answers is not in the document]
Comment (raj-prince) [reply internal/cache/lrucache/lrucache.go:37]: As discussed offline, will let the client handle to create an infinite size cache. Client needs to explicitly pass MAX_uint64 to create infinite size cache.

**Comment by raj-prince** [comment]
> Given this package will be used at multiple places, can we check we can write tests to check concurrency of exposed public methods ?



> Given this package will be used at multiple places, can we check we can write tests to check concurrency of exposed public methods ?

Went through with posts - https://stackoverflow.com/questions/19189542/how-to-test-concurrency-and-locking-in-golang, which mentions this is hard to test. This should be covered mostly by design and code review.

For now, I have added test to catch  race condition, but to automate this. Either we need to run our ci linux-test to run with `-race` flag. 

Running all the test with `race` flag is also not good. So, we need to mark individually test to not run with `race` flag. This required changes in all the package-test. 

For now I test with -race tag it is running successfully. I tried by commenting out the lock, it fails.

```
==================
[       OK ] CacheTest.TestRaceCondition
[----------] Finished with tests from CacheTest
    testing.go:1465: race detected during execution of test
--- FAIL: TestCache (0.00s)

```

Please let me know, what's your view?

File: internal/cache/lru/lrucache_test.go [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+	DataSize uint64
+}
+
+func (td testData) Size() uint64 {
+	return td.DataSize
+}
+
+// insertAndAssert inserts the given key,value in the cache and assert based on
+// the expected eviction and error.
+func (t *CacheTest) insertAndAssert(key string, val lru.ValueType, expectedEviction int, evictedValues []int64, expectedError error) {
```
Comment (sethiay) [inline internal/cache/lru/lrucache_test.go:84]: can we skip using expectedEviction as length can be determined using evictedValues?

File: internal/cache/lru/lrucache_test.go [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+// This will detect race if we run the test with `-race` flag.
+// We get the race condition failure if we remove lock from Insert or Erase method.
+func (t *CacheTest) TestRaceCondition() {
+	var wg sync.WaitGroup
+	wg.Add(3)
+
+	go func() {
+		defer wg.Done()
+		for i := 0; i < MaxSize; i++ {
```
Comment (sethiay) [inline internal/cache/lru/lrucache_test.go:214]: Is there any reason why we running till MaxSize ? I believe we just want to run for multiple iterations, is that's true then we should use different name like numIterations, count.

**Review by sethiay — COMMENTED** [review]
Left few minor comments

File: internal/cache/lru/lrucache.go [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+type entry struct {
+	Key   string
+	Value ValueType
+}
+
+// New initialize a cache with the supplied maxSize, which must be greater than
+// zero.
+func New(maxSize uint64) (c Cache) {
```
[the remark this answers is not in the document]
Comment (raj-prince) [reply internal/cache/lru/lrucache.go:75]: Done.

**Review by sethiay — APPROVED** [review]

**Pull request merged**